diff --git a/plasma/desktop/applets/tasks/tasks.cpp b/plasma/desktop/applets/tasks/tasks.cpp
index a881e16..4ebb87d 100644
--- a/plasma/desktop/applets/tasks/tasks.cpp
+++ b/plasma/desktop/applets/tasks/tasks.cpp
@@ -273,6 +273,9 @@ void Tasks::itemContextMenu(int id)
         menu->setMinimumWidth(declItem->implicitWidth());
     }
 
+    // Close menu when the delegate is destroyed
+    connect(item, SIGNAL(destroyed(QObject*)), menu, SLOT(close()));
+
     menu->exec(containment()->corona()->popupPosition(declItem, menu->size()));
     menu->deleteLater();
 }
