diff --git a/kcontrol/screensaver/screensaver.ui b/kcontrol/screensaver/screensaver.ui
index 6524e27..5cf64d6 100644
--- a/kcontrol/screensaver/screensaver.ui
+++ b/kcontrol/screensaver/screensaver.ui
@@ -21,6 +21,9 @@
       </sizepolicy>
      </property>
      <layout class="QFormLayout" name="formLayout">
+      <property name="fieldGrowthPolicy">
+       <enum>QFormLayout::ExpandingFieldsGrow</enum>
+      </property>
       <item row="0" column="1">
        <layout class="QGridLayout" name="gridLayout_2">
         <item row="0" column="0">
@@ -102,34 +105,13 @@
        </widget>
       </item>
       <item row="2" column="1">
-       <layout class="QHBoxLayout" name="horizontalLayout">
-        <item>
-         <widget class="QRadioButton" name="mPlasmaWidgetsRadio">
-          <property name="text">
-           <string>&amp;Desktop Widgets</string>
-          </property>
-         </widget>
-        </item>
-        <item>
-         <widget class="KPushButton" name="mPlasmaSetup">
-          <property name="enabled">
-           <bool>false</bool>
-          </property>
-          <property name="text">
-           <string>Configure...</string>
-          </property>
-         </widget>
-        </item>
-       </layout>
-      </item>
-      <item row="3" column="1">
        <widget class="QRadioButton" name="mScreenSaverRadio">
         <property name="text">
          <string>S&amp;creen saver</string>
         </property>
        </widget>
       </item>
-      <item row="4" column="1">
+      <item row="3" column="1">
        <widget class="QWidget" name="mScreenSaverSettings" native="true">
         <property name="enabled">
          <bool>false</bool>
@@ -292,16 +274,16 @@
   </layout>
  </widget>
  <customwidgets>
-  <customwidget>
-   <class>KIntSpinBox</class>
-   <extends>QSpinBox</extends>
-   <header>knuminput.h</header>
-  </customwidget>
   <customwidget>
    <class>KPushButton</class>
    <extends>QPushButton</extends>
    <header>kpushbutton.h</header>
   </customwidget>
+  <customwidget>
+   <class>KIntSpinBox</class>
+   <extends>QSpinBox</extends>
+   <header>knuminput.h</header>
+  </customwidget>
  </customwidgets>
  <tabstops>
   <tabstop>mEnabledCheckBox</tabstop>
@@ -309,29 +291,11 @@
   <tabstop>mLockCheckBox</tabstop>
   <tabstop>mWaitLockEdit</tabstop>
   <tabstop>mSimpleLockerRadio</tabstop>
-  <tabstop>mPlasmaWidgetsRadio</tabstop>
-  <tabstop>mPlasmaSetup</tabstop>
   <tabstop>mScreenSaverRadio</tabstop>
   <tabstop>mSaverListView</tabstop>
  </tabstops>
  <resources/>
  <connections>
-  <connection>
-   <sender>mPlasmaWidgetsRadio</sender>
-   <signal>toggled(bool)</signal>
-   <receiver>mPlasmaSetup</receiver>
-   <slot>setEnabled(bool)</slot>
-   <hints>
-    <hint type="sourcelabel">
-     <x>175</x>
-     <y>147</y>
-    </hint>
-    <hint type="destinationlabel">
-     <x>283</x>
-     <y>147</y>
-    </hint>
-   </hints>
-  </connection>
   <connection>
    <sender>mScreenSaverRadio</sender>
    <signal>toggled(bool)</signal>
diff --git a/kcontrol/screensaver/scrnsave.cpp b/kcontrol/screensaver/scrnsave.cpp
index 0125620..a9e5c4a 100644
--- a/kcontrol/screensaver/scrnsave.cpp
+++ b/kcontrol/screensaver/scrnsave.cpp
@@ -167,12 +167,8 @@ KScreenSaver::KScreenSaver(QWidget *parent, const QVariantList&)
     connect(mWaitLockEdit, SIGNAL(valueChanged(int)),
             this, SLOT(slotLockTimeoutChanged(int)));
 
-    connect(mPlasmaWidgetsRadio, SIGNAL(toggled(bool)), this, SLOT(slotEnablePlasma(bool)));
     connect(mScreenSaverRadio, SIGNAL(toggled(bool)), this, SLOT(slotEnableLegacyScreenSaver(bool)));
 
-    mPlasmaSetup->setEnabled(mPlasmaEnabled);
-    connect(mPlasmaSetup, SIGNAL(clicked()), this, SLOT(slotPlasmaSetup()));
-
     mMonitorPreview = new ScreenPreviewWidget(this);
     mMonitorPreview->setFixedSize(200,220);
     QDesktopWidget *desktop = QApplication::desktop();
@@ -300,13 +296,8 @@ void KScreenSaver::readSettings()
 
     bool legacyScreenSaver = config.readEntry("LegacySaverEnabled", false);
     mScreenSaverRadio->setChecked(legacyScreenSaver);
+
     if (!legacyScreenSaver) {
-        mPlasmaEnabled = config.readEntry("PlasmaEnabled", false);
-        mPlasmaWidgetsRadio->setChecked(mPlasmaEnabled);
-    } else {
-        mPlasmaEnabled = false;
-    }
-    if (!legacyScreenSaver && !mPlasmaEnabled) {
         mSimpleLockerRadio->setChecked(true);
     }
 
@@ -351,7 +342,6 @@ void KScreenSaver::defaults()
     slotLock( false );
     mEnabledCheckBox->setChecked(false);
     mSimpleLockerRadio->setChecked(true);
-    mPlasmaSetup->setEnabled(false);
 
     updateValues();
 
@@ -371,7 +361,6 @@ void KScreenSaver::save()
     config.writeEntry("Timeout", mTimeout);
     config.writeEntry("LockGrace", mLockTimeout);
     config.writeEntry("Lock", mLock);
-    config.writeEntry("PlasmaEnabled", mPlasmaEnabled);
     config.writeEntry("LegacySaverEnabled", mScreenSaverRadio->isChecked());
 
     if ( !mSaver.isEmpty() )
@@ -558,17 +547,6 @@ void KScreenSaver::slotEnable(bool e)
     emit changed(true);
 }
 
-void KScreenSaver::slotEnablePlasma(bool enable)
-{
-    mPlasmaEnabled = enable;
-    //FIXME even though the button's enabled, plasma isn't until the user hits apply
-    //so the button will just show the screensaver, no plasma.
-    //what should I do about this?
-    mPlasmaSetup->setEnabled(mPlasmaEnabled);
-    mChanged = true;
-    emit changed(true);
-}
-
 void KScreenSaver::slotEnableLegacyScreenSaver(bool enable)
 {
     if (mMonitor) {
@@ -578,12 +556,6 @@ void KScreenSaver::slotEnableLegacyScreenSaver(bool enable)
     emit changed(true);
 }
 
-void KScreenSaver::slotPlasmaSetup()
-{
-    org::kde::screensaver kscreensaver("org.kde.screensaver", "/ScreenSaver", QDBusConnection::sessionBus());
-    kscreensaver.setupPlasma();
-}
-
 
 //---------------------------------------------------------------------------
 //
diff --git a/kcontrol/screensaver/scrnsave.h b/kcontrol/screensaver/scrnsave.h
index 202e14e..ac2a539 100644
--- a/kcontrol/screensaver/scrnsave.h
+++ b/kcontrol/screensaver/scrnsave.h
@@ -80,9 +80,7 @@ protected Q_SLOTS:
     // new preview
     void slotPreviewExited();
     void findSavers();
-    void slotEnablePlasma(bool enable);
     void slotEnableLegacyScreenSaver(bool enable);
-    void slotPlasmaSetup();
 
 protected:
     void writeSettings();
diff --git a/ksmserver/screenlocker/ksldapp.cpp b/ksmserver/screenlocker/ksldapp.cpp
index 78c1ddf..82fb57e 100644
--- a/ksmserver/screenlocker/ksldapp.cpp
+++ b/ksmserver/screenlocker/ksldapp.cpp
@@ -161,7 +161,8 @@ void KSldApp::configure()
         m_lockGrace = -1;
     }
     m_autoLogoutTimeout = KScreenSaverSettings::autoLogout() ? KScreenSaverSettings::autoLogoutTimeout() * 1000 : 0;
-    m_plasmaEnabled = KScreenSaverSettings::plasmaEnabled();
+    // m_plasmaEnabled = KScreenSaverSettings::plasmaEnabled();
+    m_plasmaEnabled = false;
 }
 
 void KSldApp::lock()
