diff --git a/libs/plasmaclock/clockapplet.cpp b/libs/plasmaclock/clockapplet.cpp
index 482820b..18f892c 100644
--- a/libs/plasmaclock/clockapplet.cpp
+++ b/libs/plasmaclock/clockapplet.cpp
@@ -377,8 +377,9 @@ void ClockApplet::createConfigurationInterface(KConfigDialog *parent)
     if (d->kttsAvailable) {
         QWidget *generalWidget = new QWidget();
         d->generalUi.setupUi(generalWidget);
+#ifdef JOVIE_FOUND
         parent->addPage(generalWidget, i18nc("General configuration page", "General"), Applet::icon());
-
+#endif
         d->generalUi.intervalCombo->addItem(i18nc("@inmenu interval between speaking clock", "Never"), QVariant(0));
         d->generalUi.intervalCombo->addItem(i18nc("@inmenu interval between speaking clock", "Every minute"), QVariant(1));
         d->generalUi.intervalCombo->addItem(i18nc("@inmenu interval between speaking clock", "Every 2 minutes"), QVariant(2));
