diff --git a/plasma/generic/applets/systemtray/package/contents/ui/IconsList.qml b/plasma/generic/applets/systemtray/package/contents/ui/IconsList.qml
index d72733c..7ec7346 100644
--- a/plasma/generic/applets/systemtray/package/contents/ui/IconsList.qml
+++ b/plasma/generic/applets/systemtray/package/contents/ui/IconsList.qml
@@ -33,7 +33,7 @@ MouseArea {
     id: root_item
 
     property int icons_size:     24  ///< Size of icons, icons are square i.e. width == height
-    property int icons_margins:  icons_size/6  ///< Margins for icons
+    property int icons_margins:  4  ///< Margins for icons
     property alias icons_number: repeater.count  ///< [readonly] Number of icons
     property alias model:    repeater.model; ///< Model for grid
     property int cell_size: icons_size + 2*icons_margins ///< [readonly] size of grid cell
diff --git a/plasma/generic/applets/systemtray/package/contents/ui/main.qml b/plasma/generic/applets/systemtray/package/contents/ui/main.qml
index 209091b..8577362 100644
--- a/plasma/generic/applets/systemtray/package/contents/ui/main.qml
+++ b/plasma/generic/applets/systemtray/package/contents/ui/main.qml
@@ -34,7 +34,7 @@ Item {
     property int minimumWidth:  JS.MINIMUM_SIZE
     property int minimumHeight: JS.MINIMUM_SIZE
 
-    property int iconSize: Math.min(root_item.width, Math.min(root_item.height, theme.defaultFont.mSize.height < 20 ? 24 : theme.largeIconSize)) //Math.min(root_item.width, Math.min(root_item.height, JS.ICONS_SIZE))
+    property int iconSize: Math.min(root_item.width, Math.min(root_item.height, JS.ICONS_SIZE))
 
     // Data Models
     property list<ListModel> models: [
@@ -112,7 +112,7 @@ Item {
 
             content: IconsList {
                 id: popup_area
-                icons_size: root_item.iconSize
+                icons_size: JS.ICONS_SIZE
                 model: model_popup
             }
         }
