diff --git a/kwin/clients/aurorae/src/aurorae.knsrc b/kwin/clients/aurorae/src/aurorae.knsrc
index d1210b3..cf78545 100644
--- a/kwin/clients/aurorae/src/aurorae.knsrc
+++ b/kwin/clients/aurorae/src/aurorae.knsrc
@@ -1,4 +1,5 @@
 [KNewStuff3]
+ProvidersUrl=http://download.kde.org/ocs/providers.xml
 Categories=Window Decoration Aurorae
 Uncompress=archive
 TargetDir=aurorae/themes
diff --git a/kwin/kcmkwin/kwincompositing/kwineffect.knsrc b/kwin/kcmkwin/kwincompositing/kwineffect.knsrc
index e2d65ee..be71220 100644
--- a/kwin/kcmkwin/kwincompositing/kwineffect.knsrc
+++ b/kwin/kcmkwin/kwincompositing/kwineffect.knsrc
@@ -1,4 +1,5 @@
 [KNewStuff3]
+ProvidersUrl=http://download.kde.org/ocs/providers.xml
 Categories=KWin Effects
 StandardResource=tmp
 InstallationCommand=plasmapkg -t kwineffect -i %f
diff --git a/kwin/kcmkwin/kwinscripts/kwinscripts.knsrc b/kwin/kcmkwin/kwinscripts/kwinscripts.knsrc
index a223955..633568d 100644
--- a/kwin/kcmkwin/kwinscripts/kwinscripts.knsrc
+++ b/kwin/kcmkwin/kwinscripts/kwinscripts.knsrc
@@ -1,4 +1,5 @@
 [KNewStuff3]
+ProvidersUrl=http://download.kde.org/ocs/providers.xml
 Categories=KWin Scripts
 StandardResource=tmp
 InstallationCommand=plasmapkg -t kwinscript -i %f
diff --git a/kwin/kcmkwin/kwintabbox/kwinswitcher.knsrc b/kwin/kcmkwin/kwintabbox/kwinswitcher.knsrc
index da86f09..30db68e 100644
--- a/kwin/kcmkwin/kwintabbox/kwinswitcher.knsrc
+++ b/kwin/kcmkwin/kwintabbox/kwinswitcher.knsrc
@@ -1,4 +1,5 @@
 [KNewStuff3]
+ProvidersUrl=http://download.kde.org/ocs/providers.xml
 Categories=KWin Switching Layouts
 StandardResource=tmp
 InstallationCommand=plasmapkg -t windowswitcher -i %f
